File: app/lib/models/message.dart
/// A note the model drafted from an utterance ("tell mum to buy diapers"), for the app
/// to confirm and send.
class MessageDraft {
  const MessageDraft({this.to, required this.text});

  final String? to;
  final String text;

  factory MessageDraft.fromJson(Map<String, dynamic> json) => MessageDraft(
        to: json['to'] as String?,
        text: json['text'] as String,
      );
}

/// A note between caregivers, as the thread shows it.
class Message {
  const Message({
    required this.id,
    required this.text,
    this.fromName,
    this.mine = false,
    this.read = false,
    required this.createdAt,
  });

  final String id;
  final String text;
  final String? fromName;
  final bool mine;
  final bool read;
  final DateTime createdAt;

  factory Message.fromJson(Map<String, dynamic> json) => Message(
        id: json['id'] as String,
        text: json['text'] as String,
        fromName: json['from_name'] as String?,
        mine: json['mine'] as bool? ?? false,
        read: json['read'] as bool? ?? false,
        createdAt: DateTime.parse(json['created_at'] as String),
      );
}

File: server/app/models/message.py
"""Notes between the caregivers in a family.

Small and family-scoped: "tell mum to buy diapers" leaves a note the other parent
sees, with an unread badge, whether or not they were looking at the app.
"""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class MessageDraft(BaseModel):
    """A message the model drafted from an utterance ("tell mum to buy diapers").

    The app confirms it and posts it, the same way a spoken reminder rule is confirmed.
    """

    # A name the caregiver used, kept for display. The message goes to the family either
    # way -- in a two-parent family that is the other parent.
    to: Optional[str] = None
    text: str


class MessageCreate(BaseModel):
    text: str


class MessageOut(BaseModel):
    id: str
    text: str
    from_user: Optional[str] = None
    from_name: Optional[str] = None
    # True when the caller sent it, so the app can put it on the right side.
    mine: bool = False
    read: bool = False
    created_at: datetime
